mesonbuild/scripts/env2mfile.py

```diff
@@ -63,6 +63,49 @@ def __init__(self) -> None:
         self.cpu_family: T.Optional[str] = None
         self.endian: T.Optional[str] = None
 
+    @staticmethod
+    def populated(
+        compilers: T.Optional[T.Dict[str, T.List[str]]],
+        binaries: T.Optional[T.Dict[str, T.List[str]]],
+        properties: T.Optional[T.Dict[str, T.Union[str, T.List[str]]]],
+        compile_args: T.Optional[T.Dict[str, T.List[str]]],
+        link_args: T.Optional[T.Dict[str, T.List[str]]],
+        cmake: T.Optional[T.Dict[str, T.Union[str, T.List[str]]]],
+        system: T.Optional[str] = None,
+        subsystem: T.Optional[str] = None,
+        kernel: T.Optional[str] = None,
+        cpu: T.Optional[str] = None,
+        cpu_family: T.Optional[str] = None,
+        endian: T.Optional[str] = None,
+    ) -> 'MachineInfo':
+        self = MachineInfo()
+        self.system = system
+        self.subsystem = subsystem
+        self.kernel = kernel
+        self.cpu = cpu
+        self.cpu_family = cpu_family
+        self.endian = endian
+
+        if compilers is not None:
+            self.compilers.update(compilers)
+
+        if binaries is not None:
+            self.binaries.update(binaries)
+
+        if properties is not None:
+            self.properties.update(properties)
+
+        if compile_args is not None:
+            self.compile_args.update(compile_args)
+
+        if link_args is not None:
+            self.link_args.update(link_args)
+
+        if cmake is not None:
+            self.cmake.update(cmake)
+
+        return self
```

Comment:
Is there really a need for a second initializer? This really seems like we could use a `dataclass` to simplify the default initializer to be more flexible:
```python
@dataclass
class MachineInfo:
    
    compilers: T.Dict[str, T.List[str]] = field(default_factory=dict)
    binaries: T.Dict[str, T.List[str]] = field(default_factory=dict)
    ....
    system: T.Optional[str] = None
```

Reply:
I wasn't aware of dataclasses being available in Meson's minimum required Python. If this works, then yes it's probably the right way to achieve what I'm doing here.

Reply:
We have dataclasses, Python 3.7 is the current minimum supported version IIRC